### ISSUE TITLE
Mixing Two Models

### ISSUE THREAD

**Issue by laszbalo** [OHF-Voice/piper1-gpl#107]
Hi,

I tried the Hungarian female voices, Anna and Berta. They seemed fine at first, but after testing them with real text, they didn’t sound that good. Quite often, it’s difficult to understand what they’re saying.

I looked into their training data, and it appears that each was trained on just over a thousand short clips, only a few seconds long.

I know that I can train new voices, and I found the checkpoints for these models on Hugging Face. What would happen if I used the checkpoint from the first model (for example, Anna) and then retrained it using the training data from the second model (Berta)?

Would the resulting model be generally better at speaking Hungarian, but still sound like Berta? Or would it sound like a mix of both voices, perhaps with some words spoken in Anna’s voice and others in Berta’s?

**Comment by synesthesiam** [comment]
The final model would sound like Berta. Both the Anna and Berta models were fine-tuned from the U.S. English Lessac model. I'm not sure if the quality would be increased, though, as the sentences spoken by both Anna and Berta are the same. So any gaps in the coverage of certain sounds would be identical.

Can you give some specific examples of where the models are not sounding good? Maybe we could add more sentences to the training data for the future ([here are the current sentences](https://github.com/rhasspy/piper-recording-studio/tree/master/prompts/Hungarian%20(Hungary)_hu-HU)).

**Comment by laszbalo** [comment]
Thanks for the clarification.

I believe the largest Hungarian dataset is [this](https://huggingface.co/datasets/sarpba/CV22-hun-cleaned_24KHz), containing approximately 130 hours of audio. The best-quality large TTS model is based on F5-TTS, and the dataset above was used to fine-tune it. There is also another version of F5-TTS for Hungarian, based on a 12-hour dataset, which was linked in the older Piper project. That dataset originates from a single Hungarian audiobook written around 125 years ago.

So, if my understanding is correct, by using the latest checkpoint from any of the three current models and continuing training with the larger dataset mentioned above the resulting model should, in theory, speak Hungarian more fluently. Once, or if, it does, using that new checkpoint and adding back Berta’s training data should produce a model with general Hungarian language skills, but in Berta’s voice, rather than a mix of voices from the larger training dataset.

The current roughly 4-hour dataset that Piper was trained on appears to be a good mix of short and long sentences and contains words that are part of modern Hungarian usage, in contrast to texts mainly from older books. It also includes the spelling out of abbreviations and acronyms. Was this smaller dataset chosen because its sentence selection was designed to generalise better for everyday text? Would using the larger dataset add much, or would the smaller parameter size of Piper prevent it from benefiting significantly from more training data?

I understand that Piper only focuses on phoneme-to-text, here are a few examples from [this article](https://telex.hu/gazdasag/2025/10/28/lukoil-mol-benzinkutfinomito-toltoallomas), which eSpeak‑NG normalises and phonemises quite well, but where Piper generates voice that do not sound correct:

Original command: espeak-ng -v hu --ipa=3 -q → The way I would write it after hearing the spoken voice:

petrol station:
töltőállomása /tˈøltøːaːllomaːʃɑ/ → tltlomsa /tˈeːltlˈomʃɑ/

on the 28th:
28-án /hˈusonɲolt‍sɑdikaːn/ → huszolzen /hˈusolzɛn/

October:
október /ˈoktoːbɛr/ → ótber /ˈoːtbɛr/

on the 27th:
27-én /hˈusonhɛtɛdikeːn/ → huszonhéten /hˈusonheːtɛn/

arrived:
érkezett /ˈeːrkɛzɛtː/ → kezett /kˈɛzɛtː/

news:
hír /hˈiːr/ → háe /hˈaːɛ/

via:
által /ˈaːltɑl/ → tal /tˈɑl/

with suspicion:
gyanakvással /ɟˈɑnɑkvaːʃʃɑl/ → gyanaksal /ɟˈɑnɑkʃɑl/

oil company:
olajvállalat /ˈolɑjvaːllɑlɑt/ → olajvölalat /ˈolɑjvølɑlɑt/

American:
amerikai /ˈɑmɛrikɑi/ → amerikei /ˈɑmɛrikɛi/

sanctions:
szankciók /sˈɑnkt‍sioːk/ → szanciká /sˈɑnt‍sikaː/

company:
cég /t‍sˈeːɡ/ → cegé /t‍sˈɛɡeː/

stated:
közölte /kˈøzøltɛ/ → kázilte /kˈaːziltɛ/

was commenced:
megkezdődött /mˈɛkːɛzdøːdøtː/ → megkeztété /mˈɛkːɛsteːteː/

potential:
potenciális /pˈotɛnt‍siaːliʃ/ → potencilis /pˈotɛnt‍siliʃ/

buyers:
vásárlók /vˈaːʃaːrloːk/ → véserelká /vˈeːʃɛrɛlkaː/

of their offers:
ajánlatainak /ˈɑjaːllɑtɑinɑk/ → ajnlatainak /ˈɑjnlɑtɑinɑk/

assessment:
elbírálása /ˈɛlbiːraːlaːʃɑ/ → elberása /ˈɛlbɛraːʃɑ/

time:
időt /ˈidøːt/ → ídt /ˈiːtː/

United States:
Egyesült Államok /ˈɛɟːɛʃylt ˈaːllɑmok/ → Egyeselt Éllamok /ˈɛɟːɛʃɛlt ˈeːllɑmok/

That was just the first paragraph. This is where I stopped listening.

**Comment by mitrokun** [comment]
> I understand that Piper only focuses on phoneme-to-text


This part is handled by eSpeak-NG, after which the trained Piper model voices the prepared phonemes. If you want to make changes to the pronunciation, you'll have to modify the rules [dictionary](https://github.com/espeak-ng/espeak-ng/blob/master/dictsource/hu_rules). Ideally, the conversion of letters to phonemes should follow the official rules for the language, so that during the training stage, tokens can be well associated with real audio data.
You're still lucky, in Russian, you need to set the stress for almost every word to select the correct phonemes and ensure proper pronunciation. And considering how slowly the main eSpeak repository is updated, it's a real headache. You have to compile the data yourself.
As for the dataset size, it won't help if there are issues at the phoneme creation stage. Once everything is working, fine-tuning a voice can be done even with just 10 minutes of audio.

**Comment by laszbalo** [comment]
Ooop, that’s a mistake. I meant phoneme-to-speech.  For Hungarian, espeak-ng does a good job, luckily, I don’t have to recompile its dictionary, etc.. It performs text normalisation out of the box, so it converts numbers to their written forms, too. The phonemes seem fine, but the speech that comes out of Piper is off, as I tried to highlight above in response to @synesthesiam. Approximately half of the words in the first paragraph were pronounced poorly, often skipping a syllable. On two occasions, the spoken words were unrecognisable. Hence, I felt that fine-tuning with a short training dataset might not be enough.
